Adjust Engulfing Tutorial

dylanlee082 opened the pull request:
**Brief Description of What This PR Does**

This PR adds an additional check to the engulfing tutorial to only select a closest object to engulf if it is bigger than the player object size * .4. 

Hi! As this is my first time contributing I had some questions, hopefully this is the right place for them.
Questions:
- Should the .4 value be in a CONSTANT?
- I took a bit of a guess with .4. I tried using half the player objects size and found that many of the other objects in the beginning of the game are size 2 and the player object is size 5, so with .5 all of the objects are smaller than 2.5 and therefore can't be engulfed. It's also hard that the CanEngulfObject size with the ratio is typically 3. Is there a more deterministic way to come up with a value for this?
- The issue description talks about maybe using the player and cell speed to determine which objects to use, but I can't find where in the codebase speed is calculated.

**Related Issues**

closes #4916 

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

hhyyrylainen commented on src/microbe_stage/components/Engulfer.cs:
```diff
@@ -189,6 +189,10 @@ public static EngulfCheckResult CanEngulfObject(this ref Engulfer engulfer, uint
                 continue;
 
+            // Skip entities that are too small to catch easily
+            if (engulfer.EngulfingSize / .4 > engulfable.AdjustedEngulfSize)
```
Putting this here will impact AI behaviour as well. Perhaps it would be good to have a new parameter for this method to "skipLikelyTooFastTargets`? That way the info that is fed through the tutorial to the player could use that flag and the microbe AI, for now at least, could keep the old behaviour?

dylanlee082 commented:
For the tutorial it seems odd to have so much of this logic when it would potentially be easy to spawn new entities as needed for the tutorial. I looked through some of the other tutorials, like the initial glucose one, and it seems like that just isn't the pattern used so far in the game, but is that something you guys have already thought through?

hhyyrylainen commented on src/microbe_stage/components/Engulfer.cs:
```diff
@@ -189,6 +190,13 @@ public static EngulfCheckResult CanEngulfObject(this ref Engulfer engulfer, uint
+            // Skip entities that are too small to catch easily
+            if (skipLikelyTooFastTargets &&
+                engulfer.EngulfingSize / Constants.TUTORIAL_ENGULFABLE_SIZE_RATIO > engulfable.AdjustedEngulfSize)
+                {
```
Well I missed up here with the Github code editor.

I'll pull the changes locally to fix this and do some testing if the gameplay change is now good.

dylanlee082 commented:
Turns out I had a bug. My original approach on this was to divide cell size by 2 and then compare and when I realized that wasn't going to work I switched the value to .4, but then never switched from dividing to multiplying. This now fixed. I also dropped the constant from .4 to .35 to ensure it will find more entities.

hhyyrylainen reviewed (approved):
I think this is pretty well tuned now. As long as there's some engulf candidates in the first two generations this seems to always trigger. Though if the player goes for full organelle amounts for 3 generations then this tutorial might not trigger until the player adds iron organelles, but I think this is about tuned well enough to trigger for most players, I think. 

Thanks for improving this tutorial.
